src/recode.jl

```diff
@@ -96,34 +101,29 @@ function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs
     dest
 end
 
-function recode!(dest::CategoricalArray{T}, src::AbstractArray, default::Any, pairs::Pair...) where {T}
-    if length(dest) != length(src)
-        throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
-    end
-
-    opt_pairs = map(optimize_pair, pairs)
+function _recode!(dest::CategoricalArray{T, <:Any, R}, src::AbstractArray, default::Any, pairs) where {T, R}
+    recode_to = last.(pairs)
+    recode_from = first.(pairs)
 
-    vals = T[p.second for p in opt_pairs]
-    default !== nothing && push!(vals, default)
+    vals = convert.(T, recode_to)
+    vals = default === nothing ? vals : (vals..., default)
```

Comment:
Does this change really make a difference for performance? It would seem it matters more to do it for `pairmap`. It would also be worth checking that performance doesn't degrade when the tuple gets long, e.g. with 10 pairs.

Also if you do this the same change should be applied to other methods.

In terms of implementation, for type stability, don't change the type of `vals`. Also, `default` needs to be converted too.
```suggestion
    if default === nothing
        vals = convert.(T, recode_to)
    else
        vals = convert.(T, (recode_to..., default))
    end
```

Reply:
I don't really remember why I made this change so I'll just revert.